### PR TITLE
[flang] Add reduction semantics to fir.do_loop

### PR DESCRIPTION
**Pull request by khaki3**
Derived from #92480. This PR introduces reduction semantics into loops for DO CONCURRENT REDUCE. The `fir.do_loop` operation now invisibly has the `operandSegmentsizes` attribute and takes variable-length reduction operands with their operations given as `fir.reduce_attr`. For the sake of compatibility, `fir.do_loop`'s builder has additional arguments at the end. The `iter_args` operand should be placed in front of the declaration of result types, so the new operand for reduction variables (`reduce`) is put in the middle of arguments.

### CODE REVIEW

**Comment by khaki3** [comment]
I cannot find any case in which we need `fir.reduce`. The name of reduction symbols have to be declared, so we can always track their original names. I will directly put reduction variables instead.

**Comment by clementval** [comment]
> I cannot find any case in which we need `fir.reduce`. The name of reduction symbols have to be declared, so we can always track their original names. I will directly put reduction variables instead.

Yeah I think it makes sense. In OpenACC dialect with have this extra op because the dialect is meant to be language agnostic. Here we know that we deal with Fortran so it should be fine.

**Comment by khaki3** [comment]
Thanks. I removed `fir.reduce`.

File: flang/include/flang/Optimizer/Dialect/FIROps.td [hunk 1]
```diff
@@ -2138,9 +2138,11 @@ def fir_DoLoopOp : region_Op<"do_loop",
     Index:$lowerBound,
     Index:$upperBound,
     Index:$step,
+    Variadic<AnyType>:$reduceOperands,
     Variadic<AnyType>:$initArgs,
```
Comment (clementval) [inline flang/include/flang/Optimizer/Dialect/FIROps.td:2142]:
```suggestion
    Variadic<AnyType>:$initArgs,
    Variadic<AnyType>:$reduceOperands,
```

initArgs is more related to the control operands so maybe it's better to keep them together.

File: flang/lib/Optimizer/Dialect/FIROps.cpp [hunk 2]
```diff
@@ -2238,6 +2285,17 @@ void fir::DoLoopOp::print(mlir::OpAsmPrinter &p) {
...
+    p << " reduce(";
+    auto attrs = getReduceAttrsAttr();
+    auto operands = getReduceOperands();
+    llvm::interleaveComma(llvm::zip(attrs, operands), p, [&](auto it) {
+      p << std::get<0>(it) << " -> " << std::get<1>(it) << " : "
+        << std::get<1>(it).getType();
+    });
+    p << ')';
+    printBlockTerminators = true;
+  }
```
Comment (clementval) [inline flang/lib/Optimizer/Dialect/FIROps.cpp:2298]: Switch position with `iter_args`

**Review by clementval — COMMENTED** [review]
It looks mostly ok to me. Just a comment about the position of the reduce list in the op.

File: flang/lib/Optimizer/Dialect/FIROps.cpp (hunk 2 above)
Comment (khaki3) [reply flang/lib/Optimizer/Dialect/FIROps.cpp:2298]: `iter_args` affects the result types (e.g. `iter_args(%arg1 = %7) -> (index, i32)`). While we do not expect `reduce` to be coupled with `iter_args`, that is still acceptable. Which one should we choose?

- `iter_args(%arg1 = %7) reduce(...) -> (index, i32)` 
- `reduce(...) iter_args(%arg1 = %7) -> (index, i32)`

Comment (clementval) [reply flang/lib/Optimizer/Dialect/FIROps.cpp:2298]: Yeah that's a good point. Let's keep it like this.

**Review by clementval — APPROVED** [review]

**Comment by khaki3** [comment]
@klausler Could you check this PR? I appreciate it!

**Comment by klausler** [comment]
No, I have no knowledge of fir.

**Pull request merged**